Metric.read should support PEP 604 syntax for `Optional` types

Currently, annotating an attribute as `Optional[str]` works, but annotating the same attribute as `str | None` results in the following error when trying to read an empty field

```py
AssertionError: Do not know how to convert string to str | None for value:
```

--- a/fgpyo/util/inspect.py
+++ b/fgpyo/util/inspect.py
@@ -159,7 +159,7 @@ def list_parser(
     cls: Type, type_: TypeAlias, parsers: Optional[Dict[type, Callable[[str], Any]]] = None
 ) -> partial:
     """
-    Returns a function that parses a stringified list into a `List` of the correct type.
+    Returns a function that parses a "stringified" list into a `List` of the correct type.
 
     Args:
         cls: the type of the class object this is being parsed for (used to get default val for

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -166,7 +166,7 @@ def _make_union_parser_worker(
     raise ValueError(f"{value} could not be parsed as any of {union}")
 
 
-def make_union_parser(union: Type[UnionType], parsers: Iterable[Callable[[str], type]]) -> partial:
+def make_union_parser(union: Type[UnionType], parsers: Iterable[Callable[[str], UnionType]]) -> partial:
     """Generates a parser function for a union type object and set of parsers for the possible
     parsers to that union type object
     """

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -112,6 +112,9 @@ def _is_optional(dtype: TypeAnnotation) -> bool:
     An optional type may be declared using three syntaxes: `Optional[T]`, `Union[T, None]`, or `T |
     None`. All of these syntaxes are supported by this function.
 
+    Type annotations may be any of `type`, `UnionType`, `types.GenericAlias`,
+    or `typing._GenericAlias`.
+
     Args:
         dtype: A type.
 
@@ -121,8 +124,6 @@ def _is_optional(dtype: TypeAnnotation) -> bool:
 
     Raises:
         TypeError: If the input is not a valid `TypeAnnotation` type.
-        Type annotations may be any of `type`, `types.UnionType`, `types.GenericAlias`,
-        or `typing._GenericAlias`.
     """
     if not isinstance(dtype, TYPE_ANNOTATION_TYPES):
         raise TypeError(f"Expected type annotation, got {type(dtype)}: {dtype}")

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -1,5 +1,6 @@
 import collections
 import inspect
+import sys
 import typing
 from enum import Enum
 from functools import partial
@@ -11,7 +12,18 @@ from typing import TypeVar
 from typing import Union
 from typing import cast
 
-UnionType = TypeVar("UnionType", bound="Union")
+from typing_extensions import TypeAlias
+
+if sys.version_info >= (3, 10):
+    from types import UnionType
+else:
+    # NB: `types.UnionType`, available since Python 3.10, is **not** a `type`, but is a class.
+    # We declare an empty class here to use in the instance checks below.
+    class UnionType:
+        pass
+
+
+# UnionType = TypeVar("UnionType", bound="Union")
 EnumType = TypeVar("EnumType", bound="Enum")
 # conceptually bound to "Literal" but that's not valid in the spec
 # see: https://peps.python.org/pep-0586/#illegal-parameters-for-literal-at-type-check-time
@@ -73,9 +85,52 @@ def is_constructible_from_str(type_: type) -> bool:
     return False
 
 
-def _is_optional(type_: type) -> bool:
-    """Returns true if type_ is optional"""
-    return typing.get_origin(type_) is Union and type(None) in typing.get_args(type_)
+# NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
+TypeAnnotation: TypeAlias = Union[type, typing._GenericAlias, UnionType]  # type: ignore[name-defined]
+"""
+A function parameter's type annotation may be any of the following:
+    1) `type`, when declaring any of the built-in Python types
+    2) `typing._GenericAlias`, when declaring generic collection types or union types using pre-PEP
+        585 and pre-PEP 604 syntax (e.g. `List[int]`, `Optional[int]`, or `Union[int, None]`)
+    3) `types.UnionType`, when declaring union types using PEP604 syntax (e.g. `int | None`)
+    4) `types.GenericAlias`, when declaring generic collection types using PEP 585 syntax (e.g.
+       `list[int]`)
+`types.GenericAlias` is a subclass of `type`, but `typing._GenericAlias` and `types.UnionType` are
+not and must be considered explicitly.
+"""
+
+# TODO When dropping support for Python 3.9, deprecate this in favor of performing instance checks
+# directly on the `TypeAnnotation` union type.
+# NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
+TYPE_ANNOTATION_TYPES = (type, typing._GenericAlias, UnionType)  # type: ignore[attr-defined]
+
+
+def _is_optional(dtype: TypeAnnotation) -> bool:
+    """Check if a type is `Optional`.
+    An optional type may be declared using three syntaxes: `Optional[T]`, `Union[T, None]`, or `T |
+    None`. All of these syntaxes is supported by this function.
+    Args:
+        dtype: A type.
+    Returns:
+        True if the type is a union type with exactly two elements, one of which is `None`.
+        False otherwise.
+    Raises:
+        TypeError: If the input is not a valid `TypeAnnotation` type (see above).
+    """
+    if not isinstance(dtype, TYPE_ANNOTATION_TYPES):
+        raise TypeError(f"Expected type annotation, got {type(dtype)}: {dtype}")
+
+    origin = typing.get_origin(dtype)
+    args = typing.get_args(dtype)
+
+    # Optional[T] has `typing.Union` as its origin, but PEP604 syntax (e.g. `int | None`) has
+    # `types.UnionType` as its origin.
+    return (
+        origin is not None
+        and (origin is Union or origin is UnionType)
+        and len(args) == 2
+        and type(None) in args
+    )
 
 
 def _make_union_parser_worker(
@@ -84,7 +139,7 @@ def _make_union_parser_worker(
     value: str,
 ) -> UnionType:
     """Worker function behind union parsing. Iterates through possible parsers for the union and
-    returns the value produced by the first parser that works. Otherwise raises an error if none
+    returns the value produced by the first parser that works. Otherwise, raises an error if none
     work"""
     # Need to do this in the case of type Optional[str], because otherwise it'll return the string
     # 'None' instead of the object None

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -23,7 +23,6 @@ else:
         pass
 
 
-# UnionType = TypeVar("UnionType", bound="Union")
 EnumType = TypeVar("EnumType", bound="Enum")
 # conceptually bound to "Literal" but that's not valid in the spec
 # see: https://peps.python.org/pep-0586/#illegal-parameters-for-literal-at-type-check-time

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -166,7 +166,9 @@ def _make_union_parser_worker(
     raise ValueError(f"{value} could not be parsed as any of {union}")
 
 
-def make_union_parser(union: Type[UnionType], parsers: Iterable[Callable[[str], UnionType]]) -> partial:
+def make_union_parser(
+    union: Type[UnionType], parsers: Iterable[Callable[[str], UnionType]]
+) -> partial:
     """Generates a parser function for a union type object and set of parsers for the possible
     parsers to that union type object
     """

--- a/fgpyo/util/types.py
+++ b/fgpyo/util/types.py
@@ -1,6 +1,7 @@
 import collections
 import inspect
 import sys
+import types
 import typing
 from enum import Enum
 from functools import partial
@@ -85,7 +86,7 @@ def is_constructible_from_str(type_: type) -> bool:
 
 
 # NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
-TypeAnnotation: TypeAlias = Union[type, typing._GenericAlias, UnionType]  # type: ignore[name-defined]
+TypeAnnotation: TypeAlias = Union[type, typing._GenericAlias, UnionType, types.GenericAlias]  # type: ignore[name-defined]
 """
 A function parameter's type annotation may be any of the following:
     1) `type`, when declaring any of the built-in Python types
@@ -101,20 +102,27 @@ not and must be considered explicitly.
 # TODO When dropping support for Python 3.9, deprecate this in favor of performing instance checks
 # directly on the `TypeAnnotation` union type.
 # NB: since `_GenericAlias` is a private attribute of the `typing` module, mypy doesn't find it
-TYPE_ANNOTATION_TYPES = (type, typing._GenericAlias, UnionType)  # type: ignore[attr-defined]
+TYPE_ANNOTATION_TYPES = (type, typing._GenericAlias, UnionType, types.GenericAlias)  # type: ignore[attr-defined]
 
 
 def _is_optional(dtype: TypeAnnotation) -> bool:
-    """Check if a type is `Optional`.
+    """
+    Check if a type is `Optional`.
+
     An optional type may be declared using three syntaxes: `Optional[T]`, `Union[T, None]`, or `T |
-    None`. All of these syntaxes is supported by this function.
+    None`. All of these syntaxes are supported by this function.
+
     Args:
         dtype: A type.
+
     Returns:
         True if the type is a union type with exactly two elements, one of which is `None`.
         False otherwise.
+
     Raises:
-        TypeError: If the input is not a valid `TypeAnnotation` type (see above).
+        TypeError: If the input is not a valid `TypeAnnotation` type.
+        Type annotations may be any of `type`, `types.UnionType`, `types.GenericAlias`,
+        or `typing._GenericAlias`.
     """
     if not isinstance(dtype, TYPE_ANNOTATION_TYPES):
         raise TypeError(f"Expected type annotation, got {type(dtype)}: {dtype}")

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -33,7 +33,24 @@ def test_is_optional(tpe: Type, expected: bool) -> None:
     assert types._is_optional(tpe) == expected
 
 
+def test_is_optional_wrong_type() -> None:
+    with pytest.raises(TypeError, match="Expected type annotation"):
+        types._is_optional(None)
+
+
 if sys.version_info >= (3, 10):
 
     def test_is_optional_python_310() -> None:
         assert types._is_optional(str | None)
+
+
+def test_make_union_parser_worker_exception() -> None:
+    class Foo:
+        pass
+
+    class Bar:
+        pass
+
+    with pytest.raises(ValueError, match="foo could not be parsed"):
+        types._make_union_parser_worker(type(Foo | Bar), parsers=[], value="foo")
+        pass

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -3,7 +3,10 @@ from typing import Iterable
 from typing import List
 from typing import Optional
 from typing import Sequence
+from typing import Type
 from typing import Union
+
+import pytest
 
 from fgpyo.util import types
 from fgpyo.util.inspect import NoneType
@@ -16,10 +19,18 @@ def test_is_listlike() -> None:
     assert not types.is_list_like(str)
 
 
-def test_is_optional() -> None:
-    assert types._is_optional(Union[str, NoneType])
-    assert types._is_optional(Optional[str])
-    assert not types._is_optional(str)
+@pytest.mark.parametrize(
+    "tpe, expected",
+    [
+        (Union[str, NoneType], True),
+        (Optional[str], True),
+        (Union[str, int], False),
+        (Union[str, int, None], False),
+        (str, False),
+    ],
+)
+def test_is_optional(tpe: Type, expected: bool) -> None:
+    assert types._is_optional(tpe) == expected
 
 
 if sys.version_info >= (3, 10):

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -17,7 +17,6 @@ def test_is_listlike() -> None:
 
 
 def test_is_optional() -> None:
-    print(type((Union[str, NoneType])))
     assert types._is_optional(Union[str, NoneType])
     assert types._is_optional(Optional[str])
     assert not types._is_optional(str)

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -1,11 +1,29 @@
+import sys
 from typing import Iterable
 from typing import List
+from typing import Optional
 from typing import Sequence
+from typing import Union
 
 from fgpyo.util import types
+from fgpyo.util.inspect import NoneType
 
 
 def test_is_listlike() -> None:
     assert types.is_list_like(List[str])
     assert types.is_list_like(Iterable[str])
     assert types.is_list_like(Sequence[str])
+    assert not types.is_list_like(str)
+
+
+def test_is_optional() -> None:
+    print(type((Union[str, NoneType])))
+    assert types._is_optional(Union[str, NoneType])
+    assert types._is_optional(Optional[str])
+    assert not types._is_optional(str)
+
+
+if sys.version_info >= (3, 10):
+
+    def test_is_optional_python_310() -> None:
+        assert types._is_optional(str | None)

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -43,15 +43,15 @@ if sys.version_info >= (3, 10):
     def test_is_optional_python_310() -> None:
         assert types._is_optional(str | None)
 
+    def test_make_union_parser_worker_exception() -> None:
+        class Foo:
+            pass
 
-def test_make_union_parser_worker_exception() -> None:
-    class Foo:
-        pass
+        class Bar:
+            pass
 
-    class Bar:
-        pass
+        with pytest.raises(ValueError, match="foo could not be parsed"):
+            type_ = type((Foo | Bar))
+            types._make_union_parser_worker(union=type_, parsers=[], value="foo")
+            pass
 
-    with pytest.raises(ValueError, match="foo could not be parsed"):
-        type_ = type((Foo | Bar))
-        types._make_union_parser_worker(union=type_, parsers=[], value="foo")
-        pass

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -51,7 +51,6 @@ def test_make_union_parser_worker_exception() -> None:
     class Bar:
         pass
 
-
     with pytest.raises(ValueError, match="foo could not be parsed"):
         type_ = type((Foo | Bar))
         types._make_union_parser_worker(union=type_, parsers=[], value="foo")

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -51,6 +51,8 @@ def test_make_union_parser_worker_exception() -> None:
     class Bar:
         pass
 
+
     with pytest.raises(ValueError, match="foo could not be parsed"):
-        types._make_union_parser_worker(type(Foo | Bar), parsers=[], value="foo")
+        type_ = type((Foo | Bar))
+        types._make_union_parser_worker(union=type_, parsers=[], value="foo")
         pass

--- a/tests/fgpyo/util/test_types.py
+++ b/tests/fgpyo/util/test_types.py
@@ -53,5 +53,3 @@ if sys.version_info >= (3, 10):
         with pytest.raises(ValueError, match="foo could not be parsed"):
             type_ = type((Foo | Bar))
             types._make_union_parser_worker(union=type_, parsers=[], value="foo")
-            pass
-

